core/internal/filetransfer/file_transfer_azure.go

```diff
@@ -0,0 +1,416 @@
+	// but not guaranteed, that files in an artifact will share the same account
+	// URL, so we set up each account client once and reuse it when possible.
+	clients sync.Map
+
+	// once is a map of account URLs to sync.Once objects to ensure
+	// that we only set up each Azure account client once
+	once sync.Map
+}
+
+func NewAzureClientsMap() *AzureClientsMap {
+	return &AzureClientsMap{clients: sync.Map{}, once: sync.Map{}}
+}
+
+func (am *AzureClientsMap) GetAccountClient(accountUrl string) (*azblob.Client, error) {
+	client, ok := am.clients.Load(accountUrl)
+	if !ok {
+		return nil, fmt.Errorf("client not found")
+	}
+	accountClient, ok := client.(*azblob.Client)
+	if !ok {
+		return nil, fmt.Errorf("client is not an account client")
+	}
+	return accountClient, nil
+}
+
+func (am *AzureClientsMap) GetContainerClient(accountUrl string) (*container.Client, error) {
+	client, ok := am.clients.Load(accountUrl)
+	if !ok {
+		return nil, fmt.Errorf("client not found")
+	}
+	containerClient, ok := client.(*container.Client)
+	if !ok {
+		return nil, fmt.Errorf("client is not a container client")
+	}
+	return containerClient, nil
+}
+
+// AzureFileTransfer uploads or downloads files to/from Azure.
+type AzureFileTransfer struct {
+	// logger is the logger for the file transfer
+	logger *observability.CoreLogger
+
+	// fileTransferStats is used to track upload/download progress
+	fileTransferStats FileTransferStats
+
+	// background context is used to create a reader and get the client
+	ctx context.Context
+
+	// clients is a map of account URLs to azblob.Client objects
+	clients *AzureClientsMap
+
+	// containerClients is a map of container names to container.Client objects
+	containerClients *AzureClientsMap
+}
+
+// NewAzureFileTransfer creates a new fileTransfer.
+func NewAzureFileTransfer(
+	clients *AzureClientsMap,
+	logger *observability.CoreLogger,
+	fileTransferStats FileTransferStats,
+) *AzureFileTransfer {
+	ctx := context.Background()
+	if clients == nil {
+		clients = NewAzureClientsMap()
+	}
+	return &AzureFileTransfer{
+		logger:            logger,
+		fileTransferStats: fileTransferStats,
+		ctx:               ctx,
+		clients:           clients,
+		containerClients:  NewAzureClientsMap(),
+	}
+}
+
+// SetupClient sets up the Azure account client if it is not currently set.
+func (ft *AzureFileTransfer) SetupClient(accountUrl string) {
+	onceVal, _ := ft.clients.once.LoadOrStore(accountUrl, &sync.Once{})
+	once := onceVal.(*sync.Once)
+	once.Do(func() {
+		cred, err := azidentity.NewDefaultAzureCredential(nil)
+		if err != nil {
+			ft.logger.Error("Unable to create Azure credential", "err", err)
+			return
+		}
+		client, err := azblob.NewClient(accountUrl, cred, nil)
+		if err != nil {
+			ft.logger.Error("Unable to create Azure client", "err", err)
+			return
+		}
+		ft.clients.clients.Store(accountUrl, client)
+	})
+}
+
+// SetupBlobClient sets up the Azure blob client.
+func (ft *AzureFileTransfer) SetupBlobClient(
```

Comment:
it doesn't see like this function needs to be part of `AzureFileTransfer`. the only usage of `ft` is the logger and this can be done in the call site